test/topic_based_robot.py

```diff
@@ -0,0 +1,132 @@
+#      contributors may be used to endorse or promote products derived from
+#      this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+
+from collections import OrderedDict
+
+import numpy as np
+import rclpy
+from rclpy.callback_groups import ReentrantCallbackGroup
+from rclpy.node import Node
+from rclpy.qos import QoSProfile, qos_profile_sensor_data, qos_profile_system_default
+from sensor_msgs.msg import JointState
+
+
+class TopicBasedRobot(Node):
+    """Topic Based Robot to test topic_based_ros2_control interface"""
+
+    def __init__(self, joint_names: list[str]) -> None:
+        super().__init__("topic_based_robot")
+        self.joint_names = joint_names.copy()
+        self.last_joint_command = []
+        self.current_joint_state = []
+        self.callback_group = ReentrantCallbackGroup()
+        # Publisher for the robot internal joint state (Ground truth)
+        self.actual_joint_state_publisher = self.create_publisher(
+            JointState,
+            "topic_based_joint_states",
+            qos_profile=qos_profile_sensor_data,
+            callback_group=self.callback_group,
+        )
+        # Subscriber for the desired joint state from the controller
+        self.desired_joint_state_subscriber = self.create_subscription(
+            JointState,
+            "topic_based_joint_commands",
+            self.command_callback,
+            QoSProfile(depth=1),
+            callback_group=self.callback_group,
+        )
+        # Reported joint state from ros2_control
+        self.current_joint_state_subscriber = self.create_subscription(
+            JointState,
+            "joint_states",
+            self.joint_states_callback,
+            qos_profile_system_default,
+            callback_group=self.callback_group,
+        )
+
+    def filter_joint_state_msg(self, msg: JointState):
+        joint_states = []
+        for joint_name in self.joint_names:
+            try:
+                index = msg.name.index(joint_name)
+            except ValueError:
+                msg = f"Joint name '{joint_name}' not in input keys {msg.name}"
+                raise ValueError(msg) from None
```

Comment:
Should this be masking the function variable `msg`? I also don't see any try statements around the call to `filter_joint_state_msg` so what happens to a test when this is triggered?